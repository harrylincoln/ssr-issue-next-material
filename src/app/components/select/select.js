import ErrorIcon from '../error-icon/error-icon';
import Icon from '../icon/icon';
import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import styles from './select.styles';
import { withStyles } from '@material-ui/core/styles';

export const SimpleSelect = ({
    classes,
    input,
    label,
    name,
    dataSelector,
    options,
    disabled,
    meta: { touched, error, warning },
}) => {
    const classMap = cx({
        [classes.formField]: true,
        [classes.error]: touched && (error || warning),
    });
    return (
        <div className={classMap}>
            <label htmlFor={name}>
                <span className={classes.formLabel}>{label}:</span>
                <span className={classes.selectGroup}>
                    <select
                        data-selector={dataSelector}
                        disabled={disabled}
                        name={name}
                        {...input}
                    >
                        {options.map(({ value, optionLabel }) => (
                            <option key={value} value={value}>
                                {optionLabel}
                            </option>
                        ))}
                    </select>
                    <Icon name="arrowDown" />
                </span>
            </label>
            {touched &&
                ((error && (
                    <span className={classes.formWarning}>
                        <ErrorIcon />
                        {error}
                    </span>
                )) ||
                    (warning && <span className="warning">{warning}</span>))}
        </div>
    );
};

SimpleSelect.propTypes = {
    classes: PropTypes.shape({}),
    input: PropTypes.shape(),
    disabled: PropTypes.bool,
    label: PropTypes.string,
    name: PropTypes.string,
    dataSelector: PropTypes.string,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            value: PropTypes.string.isRequired,
            optionLabel: PropTypes.string.isRequired,
        })
    ).isRequired,
    meta: PropTypes.shape({
        touched: PropTypes.bool,
        warning: PropTypes.string,
        error: PropTypes.string,
    }),
};
SimpleSelect.defaultProps = {
    classes: {},
    input: {},
    disabled: false,
    name: '',
    dataSelector: '',
    label: '',
    meta: {},
};

export default withStyles(styles)(SimpleSelect);
