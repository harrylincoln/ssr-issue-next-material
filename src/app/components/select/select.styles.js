const selectStyles = (theme) => ({
    ...theme.pallete.primary.atomics,
    ...theme.pallete.primary.forms,
    ...theme.pallete.primary.generalSelectorClasses,
    selectGroup: {
        position: 'relative',
        display: 'block',
        '& select': {
            ...theme.pallete.primary.forms.formElement,
            padding: '1rem',
            appearance: 'none',
        },
        '& svg': {
            position: 'absolute',
            right: '1rem',
            top: '1rem',
            pointerEvents: 'none',
        },
    },

    error: {
        '& select': {
            ...theme.pallete.primary.forms.formElementError,
        },
    },
});

export default selectStyles;
