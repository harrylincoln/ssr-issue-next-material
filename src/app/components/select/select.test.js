import ErrorIcon from '../error-icon/error-icon';
import React from 'react';
import { SimpleSelect as Select } from './select';
import { shallow } from 'enzyme';

describe('Select', () => {
    const error = 'error';
    const warning = 'warning';
    const touched = true;
    const options = [
        { value: '', optionLabel: 'Select' },
        { value: 'mr', optionLabel: 'Mr.' },
        { value: 'ms', optionLabel: 'Ms.' },
        { value: 'sir', optionLabel: 'Sir' },
        { value: 'mrs', optionLabel: 'Mrs.' },
        { value: 'miss', optionLabel: 'Miss.' },
        { value: 'dr', optionLabel: 'Dr.' },
    ];

    const spyScrollTo = jest.fn();
    beforeEach(() => {
        Object.defineProperty(global.window, 'scrollTo', {
            value: spyScrollTo,
        });
        Object.defineProperty(global.window, 'scrollY', { value: 1 });
        spyScrollTo.mockClear();
    });

    it('should render correctly', () => {
        const component = shallow(
            <Select name="name" label="label" options={options} />
        );
        expect(component).toMatchSnapshot();
    });

    it('should render error message', () => {
        const component = shallow(
            <Select
                meta={{ error, touched }}
                name="name"
                label="label"
                options={options}
                classes={{ formWarning: 'formWarning' }}
            />
        );

        expect(
            component.contains(
                <span className="formWarning">
                    <ErrorIcon />
                    {error}
                </span>
            )
        ).toEqual(true);
    });

    it('should render warning', () => {
        const component = shallow(
            <Select
                meta={{ warning, touched }}
                name="name"
                label="label"
                options={options}
            />
        );
        expect(
            component.contains(<span className="warning">{warning}</span>)
        ).toEqual(true);
    });
    describe('select callbacks', () => {
        const input = {
            onChange: jest.fn(),
            onBlur: jest.fn(),
            onFocus: jest.fn(),
        };

        afterEach(() => {
            input.onChange.mockClear();
            input.onBlur.mockClear();
            input.onFocus.mockClear();
        });

        it('should call onFocus for focus event', () => {
            const component = shallow(
                <Select
                    name="name"
                    label="label"
                    options={options}
                    input={input}
                />
            );
            const select = component.find('select');

            select.simulate('focus');
            expect(input.onFocus).toHaveBeenCalledTimes(1);
        });
        it('should call onChange for change event', () => {
            const component = shallow(
                <Select
                    name="name"
                    label="label"
                    options={options}
                    input={input}
                />
            );
            const select = component.find('select');

            select.simulate('change', { a: 'b' });
            expect(input.onChange).toHaveBeenCalledTimes(1);
            expect(input.onChange).toHaveBeenCalledWith({ a: 'b' });
        });

        it('should call onBlur for blur event', () => {
            const component = shallow(
                <Select
                    name="name"
                    label="label"
                    options={options}
                    input={input}
                />
            );
            const select = component.find('select');

            select.simulate('blur');
            expect(input.onBlur).toHaveBeenCalledTimes(1);
        });
    });
});
