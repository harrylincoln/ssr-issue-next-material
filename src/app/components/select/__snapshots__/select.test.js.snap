// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly 1`] = `
<div
  className=""
>
  <label
    htmlFor="name"
  >
    <span>
      label
      :
    </span>
    <span>
      <select
        data-selector=""
        disabled={false}
        name="name"
      >
        <option
          key=""
          value=""
        >
          Select
        </option>
        <option
          key="mr"
          value="mr"
        >
          Mr.
        </option>
        <option
          key="ms"
          value="ms"
        >
          Ms.
        </option>
        <option
          key="sir"
          value="sir"
        >
          Sir
        </option>
        <option
          key="mrs"
          value="mrs"
        >
          Mrs.
        </option>
        <option
          key="miss"
          value="miss"
        >
          Miss.
        </option>
        <option
          key="dr"
          value="dr"
        >
          Dr.
        </option>
      </select>
      <Icon
        name="arrowDown"
        size={24}
        style={null}
      />
    </span>
  </label>
</div>
`;
